File: c++/20_valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::vector<char> stack;

        for (char c : s) {
            switch (c) {
                case ')':
                    if (stack.empty() || stack.back() != '(') {
                        return false;
                    } else {
                        stack.pop_back();
                    }
                    break;
                case ']':
                    if (stack.empty() || stack.back() != '[') {
                        return false;
                    } else {
                        stack.pop_back();
                    }
                    break;
                case '}':
                    if (stack.empty() || stack.back() != '{') {
                        return false;
                    } else {
                        stack.pop_back();
                    }
                    break;
                default:
                    stack.push_back(c);
            }
        }

        return stack.empty();
    }
};

File: java/20_valid_parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> parens = new Stack();

        for (int i = 0; i < s.length(); i++) {
            char cur_char = s.charAt(i);
            switch (cur_char) {
                case '(':
                    parens.push(cur_char);
                    break;
                case '[':
                    parens.push(cur_char);
                    break;
                case '{':
                    parens.push(cur_char);
                    break;
                case ')':
                    if (parens.empty() || parens.pop() != '(') {
                        return false;
                    }    
                    break;
                case ']':
                    if (parens.empty() || parens.pop() != '[') {
                        return false;
                    }    
                    break;
                case '}':
                    if (parens.empty() || parens.pop() != '{') {
                        return false;
                    }    
                    break;
                default:
                    return false;
            }
        }

        return parens.empty();
    }
}
